getnacelle/nacelle-js
# refactor: custom fragments

## Why are these changes introduced?

Addresses [ENG-6852](https://nacelle.atlassian.net/browse/ENG-6852)

> Update README to describe custom fragments



## What is this pull request doing?

### Code changes

This PR removes the `EXTEND_CART_LINE_MERCHANDISE` fragment and replaces it with a `MERCHANDISE` fragment. This change is motivated by a desire for "replacement fragments" and "extension fragments" to be internally consistent in the way that they work.

When writing the `customFragments` docs, it became clear that `EXTEND_CART_LINE_MERCHANDISE` was an awkward edge case in that it was the only extension fragment that could result in _less_ data being returned, unless the developer took care to include all properties of the default `EXTEND_CART_LINE_MERCHANDISE` fragment.

This change makes it easier to document the behavior of replacement vs. extension fragments, and will hopefully improve clarity and reduce opportunities for unexpected issues.

### Docs

This PR adds a small callout in the README to a new `packages/shopify-cart/docs/custom-fragments.md` document, in an effort to keep the README manageable for readers.

## 

## How to Test

Note: some of the links in the new docs aren't valid yet, because they point to `main`. They will work as expected when `ENG-6843/shopify-cart-1.0.0` is merged into `main`.

1. Look at https://github.com/getnacelle/nacelle-js/tree/ENG-6852-document-custom-fragments/packages/shopify-cart#querying-additional-fields-with-custom-graphql-fragments.
2. Look at https://github.com/getnacelle/nacelle-js/blob/ENG-6852-document-custom-fragments/packages/shopify-cart/docs/custom-fragments.md#custom-fragments-in-nacelleshopify-cart.
3. Check out the `ENG-6852-document-custom-fragments` branch.
4. Navigate to `packages/shopify-cart`
5. `npm run test:ci` - all tests should pass with 100% coverage.

<img width="593" alt="ENG-6852-document-custom-fragments all tests passing" src="https://user-images.githubusercontent.com/5732000/188237631-01dbbe54-c107-41fe-9765-78d4bed2e563.png">

6. `npm run build` - the client should build without errors.

<img width="525" alt="ENG-6852-document-custom-fragments builds without errors" src="https://user-images.githubusercontent.com/5732000/188237649-eaa713eb-05fe-4f67-9b59-4c1c10734246.png">

7. `npm run codegen` - there should be no changes to `packages/shopify-cart/src/types/shopify.type.ts`.
